### PR TITLE
haskell.packages.ghc942: make it work

### PR DESCRIPTION
**Pull request by lf-**
###### Description of changes

Bring up ghc-9.4 enough that I can run `hpack` and the HLS dev shell, and HLS (with some limited overrides). It's still fairly thoroughly jacked. I copied the definitions from ghc-9.2 as a starting point then started fixing stuff.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lf-** [comment]
> Thank you for working on this! This is much needed. I am a bit weary of just carrying forward all those overrides without being sure if they are all needed/desirable. On the other hand this is probably the smart thing to do.

Me too, I'd love to factor them out into a >=9.2 file but i have no idea how to.

> Currently you are including a breaking change for callCabal2nixWithOptions in this PR was that maybe a mistake?

Only half a mistake; I needed it for a project and didn't want to deal with another branch. It's not breaking as it's backward compatible: I added `callCabal2nixWithOptions'` which lets you configure the source filter for cabal files, which is necessary to use hpack include directives, but `callCabal2nixWithOptions` remains as-is with the same API. If you'd like me to split it into another PR I can do that for you.

**Comment by sternenseemann** [comment]
> If you'd like me to split it into another PR I can do that for you.

Please do, I think it needs separate discussion.

My request would be to remove all overrides that don't actually lead to a working package. Which those are can be detected using:

```
nix-build maintainers/scripts/haskell/test-configurations.nix --keep-going --argstr files pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix
```

All failing packages are an override in the configuration, but don't build properly.

**Comment by lf-** [comment]
Reverted, going to let my computer crunch on the builds for a while on those reverts.

**Comment by lf-** [comment]
note to self for later:
```
error: build of '/nix/store/0b6znjv12cl3kipiwq7b4s7xybwghffc-ghc-source-gen-0.4.3.0.drv', '/nix/store/0hd4mzb9hr4baihj5f5d3x2577rkvy9r-regex-rure-0.1.2.0.drv', '/nix/store/0yddwkana0pv7av04bb2paim2shf5d6d-hlint-3.4.1.drv', '/nix/store/48glx2agpmdwca45c6r5b4yk1y7ppafg-servant-docs-0.12.drv', '/nix/store/4qkhwazmgqskvksjvrcyidswrhsfhf7p-servant-auth-server-0.4.7.0.drv', '/nix/store/4xdlj3cv91x6b3fbs7x5z51bxdq4bmx0-ghc-lib-parser-ex-9.2.1.1.drv', '/nix/store/6q10wnxz5hbq9yh5pk1pz8gpvc1p1cy8-servant-auth-swagger-0.2.10.1.drv', '/nix/store/awshsdbxhghwrd50rq759hvrh4az60ns-retrie-1.2.0.1.drv', '/nix/store/d8xcydixmv2x5m3g8gvynjzz1gyjcyc5-cabal-install-3.8.1.0.drv', '/nix/store/f5a7wg75f4sj0xgvza43ys2vd1hl54c9-servant-auth-docs-0.2.10.0.drv', '/nix/store/f984zfznhqlccy4cpm9f9ngh0v6my07c-tomland-1.3.3.1.drv', '/nix/store/fc8yypsh1xq3npv1hk4jhk0qgdyl32a5-tasty-hedgehog-1.1.0.0.drv', '/nix/store/fkq40aylz9jy6z2ma4nf3i8al59sw1w0-servant-auth-0.4.1.0.drv', '/nix/store/hls03ka4lm2lnxjz7a66964l9dcqf7lf-jacinda-1.1.0.0.drv', '/nix/store/m1m1vimd67s4xim2qpsymg5g30vx2iq9-ormolu-0.5.0.1.drv', '/nix/store/m7ly3c0j05ji6v9l1qf3yki4nwvs78rf-double-conversion-2.0.4.1.drv', '/nix/store/md1lfhll1gbc6xbhvc03xc6w2d3xhpn3-ghc-lib-9.2.4.20220729.drv', '/nix/store/mgp1iz5lcrfy52wqaf90qkd5j4lpbgf5-servant-swagger-1.1.11.drv', '/nix/store/nrlzk5cxfkh1i91k58i0xwsydjjdhv51-hiedb-0.4.1.0.drv', '/nix/store/nx7pacybizx0xvas9bs2ww2j8b9r5mfk-servant-multipart-0.12.1.drv', '/nix/store/pjvhx5d80pzs7qwknbsqsf5mifp13xfn-hls-fourmolu-plugin-1.0.3.0.drv', '/nix/store/pk8yi0rp8yv635cdvwkr25vv7wpf4qzs-servant-foreign-0.15.4.drv', '/nix/store/qdn7x08bgcksyg87pih1c513mkrk45mx-implicit-hie-cradle-0.5.0.0.drv', '/nix/store/sai0kx6issymhsza5kywsbfjaii7jnca-shelly-1.10.0.drv', '/nix/store/sfhqp52p1gjy218ywwhi57spwb2jbshg-hls-ormolu-plugin-1.0.2.1.drv', '/nix/store/vmlj0lh36297170nfvv2n3i7988vkli7-servant-0.19.drv', '/nix/store/w595libzbvkxjaqfr22abm6j7ka3l4pc-ghc-exactprint-1.5.0.drv', '/nix/store/wclj4pi9j8dhs47pzmx25v5kwk3idsxb-dbus-1.2.26.drv', '/nix/store/wfnzx2dg7g5r1n7d31gqllxk2hrajbl8-fourmolu-0.6.0.0.drv', '/nix/store/wr6ssw2alp3vh8kzsxjjgdn893ymd5xv-ghc-lib-parser-9.2.4.20220729.drv', '/nix/store/xb2sgw3jkh13939xa3kqm31x0nrrxdsv-doctest-0.20.0.drv', '/nix/store/zrhnrf6p320nw1dyzw1bx6y1qx1qd49z-aeson-diff-1.1.0.13.drv' failed
```

File: pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix [hunk 1]
```diff
@@ -48,4 +55,210 @@ self: super: {
...
+  haskell-src-meta = doJailbreak super.haskell-src-meta;
+
+  # Tests fail in GHC 9.2
+  extra = dontCheck super.extra;
+
+  # Jailbreaks & Version Updates
+
+  aeson = self.aeson_2_1_1_0;
+  # This `doJailbreak` can be removed once we have doctest v0.20
+  aeson-diff = assert super.doctest.version == "0.18.2"; doJailbreak super.aeson-diff;
```
Comment (ttuegel) [inline pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix:108]: Should this also be `doctest_0_20_broken`?

File: pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix [hunk 2]
```diff
@@ -48,4 +55,210 @@ self: super: {
...
+  # need bytestring >= 0.11 which is only bundled with GHC >= 9.2
+  regex-rure = doDistribute (markUnbroken super.regex-rure);
+  jacinda = doDistribute super.jacinda;
+  some = doJailbreak super.some;
+
+  fourmolu = super.fourmolu_0_8_2_0;
+  # hls-fourmolu-plugin in this version has a to strict upper bound of fourmolu <= 0.5.0.0
+  # hls-fourmolu-plugin = assert super.hls-fourmolu-plugin.version == "1.0.3.0"; doJailbreak super.hls-fourmolu-plugin;
+
+  # hls-ormolu-plugin = assert super.hls-ormolu-plugin.version == "1.0.2.1"; doJailbreak super.hls-ormolu-plugin;
```
Comment (ttuegel) [inline pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix:257]: If we don't need these, can we remove them?

Comment (9999years) [reply pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix]: done

File: pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix (hunk 1 above)
Comment (lf-) [reply pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix:108]: The reason I left this line in is that doctest 0.20 out but it's *itself* broken.

**Comment by sternenseemann** [comment]
`tasty-hspec` seems to hit a possible Cabal bug:

```
setupCompilerEnvironmentPhase
Build with /nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2.
unpacking sources
unpacking source archive /nix/store/7qdf9509d4qryj35n97cwrv9n43z5939-tasty-hspec-1.2.0.1.tar.gz
source root is tasty-hspec-1.2.0.1
setting SOURCE_DATE_EPOCH to timestamp 1000000000 of file tasty-hspec-1.2.0.1/tasty-hspec.cabal
patching sources
Replace Cabal file with edited version from mirror://hackage/tasty-hspec-1.2.0.1/revision/1.cabal.
compileBuildDriverPhase
setupCompileFlags: -package-db=/build/tmp.AzzXF5C2Nn/setup-package.conf.d -j4 +RTS -A64M -RTS -threaded -rtsopts
[1 of 2] Compiling Main             ( Setup.hs, /build/tmp.AzzXF5C2Nn/Main.o )
[2 of 2] Linking Setup
configuring
configureFlags: --verbose --prefix=/nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1 --libdir=$prefix/lib/$compiler --libsubdir=$abi/$libname --docdir=/nix/store/b5w8r3bc5vwmk4kx5jqfjhd20visrcbz-tasty-hspec-1.2.0.1-doc/share/doc/tasty-hspec-1.2.0.1 --with-gcc=gcc --package-db=/build/tmp.AzzXF5C2Nn/package.conf.d --ghc-options=-j4 +RTS -A64M -RTS --disable-split-objs --enable-library-profiling --profiling-detail=exported-functions --disable-profiling --enable-shared --disable-coverage --enable-static --disable-executable-dynamic --enable-tests --disable-benchmarks --enable-library-vanilla --disable-library-for-ghci --ghc-option=-split-sections --ghc-options=-haddock --extra-lib-dirs=/nix/store/fw7wl6m3jc2zcrj5zq794xcpb281dnh5-ncurses-6.3-p20220507/lib --extra-lib-dirs=/nix/store/3cac527kk6fsgvxgf1x7v0r1mg2j42ng-libffi-3.4.2/lib --extra-lib-dirs=/nix/store/gdrxvhnf6zh41r861a86zvcv0w0cy0rp-gmp-with-cxx-6.2.1/lib --extra-lib-dirs=/nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/lib
Using Parsec parser
Configuring tasty-hspec-1.2.0.1...
Dependency QuickCheck >=2.7 && <2.15: using QuickCheck-2.14.2
Dependency base >=4.9 && <4.18: using base-4.17.0.0
Dependency hspec >=2.5 && <2.11: using hspec-2.10.6
Dependency hspec-core >=2.5 && <2.11: using hspec-core-2.10.6
Dependency tasty >=0.8 && <0.13 || >=1.0 && <1.5: using tasty-1.4.2.3
Dependency tasty-quickcheck >=0.3 && <0.9 || >=0.9.1 && <0.11: using
tasty-quickcheck-0.10.2
Dependency tasty-smallcheck >=0.1 && <0.9: using tasty-smallcheck-0.8.2
Source component graph: component lib
Configured component graph:
    component tasty-hspec-1.2.0.1-EYPyYWfgkHoGHjMrUz3frm
        include QuickCheck-2.14.2-3tClwZe1MDbLIlIgpxfKkC
        include base-4.17.0.0
        include hspec-2.10.6-Fl3P4tqYSgk4Pft4nVrVI4
        include hspec-core-2.10.6-93d1KMsgtAZL9XXcrjnZoK
        include tasty-1.4.2.3-KylMI6VfqDu6zatv8M3Rf4
        include tasty-quickcheck-0.10.2-5d6Mmyqw94Q2ygyHB68mwL
        include tasty-smallcheck-0.8.2-LOkXEKlXvkjGHGbsJBa02w
Linked component graph:
    unit tasty-hspec-1.2.0.1-EYPyYWfgkHoGHjMrUz3frm
        include QuickCheck-2.14.2-3tClwZe1MDbLIlIgpxfKkC
        include base-4.17.0.0
        include hspec-2.10.6-Fl3P4tqYSgk4Pft4nVrVI4
        include hspec-core-2.10.6-93d1KMsgtAZL9XXcrjnZoK
        include tasty-1.4.2.3-KylMI6VfqDu6zatv8M3Rf4
        include tasty-quickcheck-0.10.2-5d6Mmyqw94Q2ygyHB68mwL
        include tasty-smallcheck-0.8.2-LOkXEKlXvkjGHGbsJBa02w
        Test.Tasty.Hspec=tasty-hspec-1.2.0.1-EYPyYWfgkHoGHjMrUz3frm:Test.Tasty.Hspec
Ready component graph:
    definite tasty-hspec-1.2.0.1-EYPyYWfgkHoGHjMrUz3frm
        depends QuickCheck-2.14.2-3tClwZe1MDbLIlIgpxfKkC
        depends base-4.17.0.0
        depends hspec-2.10.6-Fl3P4tqYSgk4Pft4nVrVI4
        depends hspec-core-2.10.6-93d1KMsgtAZL9XXcrjnZoK
        depends tasty-1.4.2.3-KylMI6VfqDu6zatv8M3Rf4
        depends tasty-quickcheck-0.10.2-5d6Mmyqw94Q2ygyHB68mwL
        depends tasty-smallcheck-0.8.2-LOkXEKlXvkjGHGbsJBa02w
Warning:
    This package indirectly depends on multiple versions of the same package. This is very likely to cause a compile failure.
      package hspec (hspec-2.10.6-Fl3P4tqYSgk4Pft4nVrVI4) requires hspec-core-2.10.6-1YaOpTrOnZw95xaaCyezQr
      package tasty-hspec (tasty-hspec-1.2.0.1) requires hspec-core-2.10.6-93d1KMsgtAZL9XXcrjnZoK
Using Cabal-3.8.1.0 compiled by ghc-9.4
Using compiler: ghc-9.4.2
Using install prefix:
/nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1
Executables installed in:
/nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/bin
Libraries installed in:
/nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/lib/ghc-9.4.2/x86_64-linux-ghc-9.4.2/tasty-hspec-1.2.0.1-EYPyYWfgkHoGHjMrUz3frm
Dynamic Libraries installed in:
/nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/lib/ghc-9.4.2/x86_64-linux-ghc-9.4.2
Private executables installed in:
/nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/libexec/x86_64-linux-ghc-9.4.2/tasty-hspec-1.2.0.1
Data files installed in:
/nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/share/x86_64-linux-ghc-9.4.2/tasty-hspec-1.2.0.1
Documentation installed in:
/nix/store/b5w8r3bc5vwmk4kx5jqfjhd20visrcbz-tasty-hspec-1.2.0.1-doc/share/doc/tasty-hspec-1.2.0.1
Configuration files installed in:
/nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/etc
No alex found
Using ar found on system at:
/nix/store/mx37gfqvx3kh8icib3k9jhvw8pv66ryv-binutils-2.39/bin/ar
No c2hs found
No cpphs found
No doctest found
Using gcc version 11.3.0 given by user at:
/nix/store/j061mvdxw70xz20r2r0lr5lk0rcsjf9i-gcc-wrapper-11.3.0/bin/gcc
Using ghc version 9.4.2 found on system at:
/nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/ghc
Using ghc-pkg version 9.4.2 found on system at:
/nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/ghc-pkg-9.4.2
No ghcjs found
No ghcjs-pkg found
No greencard found
Using haddock version 2.27.0 found on system at:
/nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/haddock-ghc-9.4.2
No happy found
Using haskell-suite found on system at: haskell-suite-dummy-location
Using haskell-suite-pkg found on system at: haskell-suite-pkg-dummy-location
No hmake found
Using hpc version 0.68 found on system at:
/nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/hpc
Using hsc2hs version 0.68.8 found on system at:
/nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/hsc2hs
Using hscolour version 1.24 found on system at:
/nix/store/9br4bs74rvsl039lpq65lxqq6v59vmyc-hscolour-1.24.4/bin/HsColour
No jhc found
Using ld found on system at:
/nix/store/hk75004j435l09ij4gawlkzg4wfs839w-binutils-wrapper-2.39/bin/ld.gold
No pkg-config found
Using runghc version 9.4.2 found on system at:
/nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/runghc-9.4.2
Using strip version 2.39 found on system at:
/nix/store/j061mvdxw70xz20r2r0lr5lk0rcsjf9i-gcc-wrapper-11.3.0/bin/strip
Using tar found on system at:
/nix/store/gb4zp7iygbkr3bqfv9p38q3ci04m4m5s-gnutar-1.34/bin/tar
No uhc found
*** abort because of serious configure-time warning from Cabal
```

The additional `hspec-core-2.10.6-1YaOpTrOnZw95xaaCyezQr` it complains about doesn't seem to actually exist.

**Comment by lf-** [comment]
> `tasty-hspec` seems to hit a possible Cabal bug:
> 
> ```
> setupCompilerEnvironmentPhase
> Build with /nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2.
> unpacking sources
> unpacking source archive /nix/store/7qdf9509d4qryj35n97cwrv9n43z5939-tasty-hspec-1.2.0.1.tar.gz
> source root is tasty-hspec-1.2.0.1
> setting SOURCE_DATE_EPOCH to timestamp 1000000000 of file tasty-hspec-1.2.0.1/tasty-hspec.cabal
> patching sources
> Replace Cabal file with edited version from mirror://hackage/tasty-hspec-1.2.0.1/revision/1.cabal.
> compileBuildDriverPhase
> setupCompileFlags: -package-db=/build/tmp.AzzXF5C2Nn/setup-package.conf.d -j4 +RTS -A64M -RTS -threaded -rtsopts
> [1 of 2] Compiling Main             ( Setup.hs, /build/tmp.AzzXF5C2Nn/Main.o )
> [2 of 2] Linking Setup
> configuring
> configureFlags: --verbose --prefix=/nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1 --libdir=$prefix/lib/$compiler --libsubdir=$abi/$libname --docdir=/nix/store/b5w8r3bc5vwmk4kx5jqfjhd20visrcbz-tasty-hspec-1.2.0.1-doc/share/doc/tasty-hspec-1.2.0.1 --with-gcc=gcc --package-db=/build/tmp.AzzXF5C2Nn/package.conf.d --ghc-options=-j4 +RTS -A64M -RTS --disable-split-objs --enable-library-profiling --profiling-detail=exported-functions --disable-profiling --enable-shared --disable-coverage --enable-static --disable-executable-dynamic --enable-tests --disable-benchmarks --enable-library-vanilla --disable-library-for-ghci --ghc-option=-split-sections --ghc-options=-haddock --extra-lib-dirs=/nix/store/fw7wl6m3jc2zcrj5zq794xcpb281dnh5-ncurses-6.3-p20220507/lib --extra-lib-dirs=/nix/store/3cac527kk6fsgvxgf1x7v0r1mg2j42ng-libffi-3.4.2/lib --extra-lib-dirs=/nix/store/gdrxvhnf6zh41r861a86zvcv0w0cy0rp-gmp-with-cxx-6.2.1/lib --extra-lib-dirs=/nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/lib
> Using Parsec parser
> Configuring tasty-hspec-1.2.0.1...
> Dependency QuickCheck >=2.7 && <2.15: using QuickCheck-2.14.2
> Dependency base >=4.9 && <4.18: using base-4.17.0.0
> Dependency hspec >=2.5 && <2.11: using hspec-2.10.6
> Dependency hspec-core >=2.5 && <2.11: using hspec-core-2.10.6
> Dependency tasty >=0.8 && <0.13 || >=1.0 && <1.5: using tasty-1.4.2.3
> Dependency tasty-quickcheck >=0.3 && <0.9 || >=0.9.1 && <0.11: using
> tasty-quickcheck-0.10.2
> Dependency tasty-smallcheck >=0.1 && <0.9: using tasty-smallcheck-0.8.2
> Source component graph: component lib
> Configured component graph:
>     component tasty-hspec-1.2.0.1-EYPyYWfgkHoGHjMrUz3frm
>         include QuickCheck-2.14.2-3tClwZe1MDbLIlIgpxfKkC
>         include base-4.17.0.0
>         include hspec-2.10.6-Fl3P4tqYSgk4Pft4nVrVI4
>         include hspec-core-2.10.6-93d1KMsgtAZL9XXcrjnZoK
>         include tasty-1.4.2.3-KylMI6VfqDu6zatv8M3Rf4
>         include tasty-quickcheck-0.10.2-5d6Mmyqw94Q2ygyHB68mwL
>         include tasty-smallcheck-0.8.2-LOkXEKlXvkjGHGbsJBa02w
> Linked component graph:
>     unit tasty-hspec-1.2.0.1-EYPyYWfgkHoGHjMrUz3frm
>         include QuickCheck-2.14.2-3tClwZe1MDbLIlIgpxfKkC
>         include base-4.17.0.0
>         include hspec-2.10.6-Fl3P4tqYSgk4Pft4nVrVI4
>         include hspec-core-2.10.6-93d1KMsgtAZL9XXcrjnZoK
>         include tasty-1.4.2.3-KylMI6VfqDu6zatv8M3Rf4
>         include tasty-quickcheck-0.10.2-5d6Mmyqw94Q2ygyHB68mwL
>         include tasty-smallcheck-0.8.2-LOkXEKlXvkjGHGbsJBa02w
>         Test.Tasty.Hspec=tasty-hspec-1.2.0.1-EYPyYWfgkHoGHjMrUz3frm:Test.Tasty.Hspec
> Ready component graph:
>     definite tasty-hspec-1.2.0.1-EYPyYWfgkHoGHjMrUz3frm
>         depends QuickCheck-2.14.2-3tClwZe1MDbLIlIgpxfKkC
>         depends base-4.17.0.0
>         depends hspec-2.10.6-Fl3P4tqYSgk4Pft4nVrVI4
>         depends hspec-core-2.10.6-93d1KMsgtAZL9XXcrjnZoK
>         depends tasty-1.4.2.3-KylMI6VfqDu6zatv8M3Rf4
>         depends tasty-quickcheck-0.10.2-5d6Mmyqw94Q2ygyHB68mwL
>         depends tasty-smallcheck-0.8.2-LOkXEKlXvkjGHGbsJBa02w
> Warning:
>     This package indirectly depends on multiple versions of the same package. This is very likely to cause a compile failure.
>       package hspec (hspec-2.10.6-Fl3P4tqYSgk4Pft4nVrVI4) requires hspec-core-2.10.6-1YaOpTrOnZw95xaaCyezQr
>       package tasty-hspec (tasty-hspec-1.2.0.1) requires hspec-core-2.10.6-93d1KMsgtAZL9XXcrjnZoK
> Using Cabal-3.8.1.0 compiled by ghc-9.4
> Using compiler: ghc-9.4.2
> Using install prefix:
> /nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1
> Executables installed in:
> /nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/bin
> Libraries installed in:
> /nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/lib/ghc-9.4.2/x86_64-linux-ghc-9.4.2/tasty-hspec-1.2.0.1-EYPyYWfgkHoGHjMrUz3frm
> Dynamic Libraries installed in:
> /nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/lib/ghc-9.4.2/x86_64-linux-ghc-9.4.2
> Private executables installed in:
> /nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/libexec/x86_64-linux-ghc-9.4.2/tasty-hspec-1.2.0.1
> Data files installed in:
> /nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/share/x86_64-linux-ghc-9.4.2/tasty-hspec-1.2.0.1
> Documentation installed in:
> /nix/store/b5w8r3bc5vwmk4kx5jqfjhd20visrcbz-tasty-hspec-1.2.0.1-doc/share/doc/tasty-hspec-1.2.0.1
> Configuration files installed in:
> /nix/store/kv1cjssrvgli5ks2ym8rgghal588ba82-tasty-hspec-1.2.0.1/etc
> No alex found
> Using ar found on system at:
> /nix/store/mx37gfqvx3kh8icib3k9jhvw8pv66ryv-binutils-2.39/bin/ar
> No c2hs found
> No cpphs found
> No doctest found
> Using gcc version 11.3.0 given by user at:
> /nix/store/j061mvdxw70xz20r2r0lr5lk0rcsjf9i-gcc-wrapper-11.3.0/bin/gcc
> Using ghc version 9.4.2 found on system at:
> /nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/ghc
> Using ghc-pkg version 9.4.2 found on system at:
> /nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/ghc-pkg-9.4.2
> No ghcjs found
> No ghcjs-pkg found
> No greencard found
> Using haddock version 2.27.0 found on system at:
> /nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/haddock-ghc-9.4.2
> No happy found
> Using haskell-suite found on system at: haskell-suite-dummy-location
> Using haskell-suite-pkg found on system at: haskell-suite-pkg-dummy-location
> No hmake found
> Using hpc version 0.68 found on system at:
> /nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/hpc
> Using hsc2hs version 0.68.8 found on system at:
> /nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/hsc2hs
> Using hscolour version 1.24 found on system at:
> /nix/store/9br4bs74rvsl039lpq65lxqq6v59vmyc-hscolour-1.24.4/bin/HsColour
> No jhc found
> Using ld found on system at:
> /nix/store/hk75004j435l09ij4gawlkzg4wfs839w-binutils-wrapper-2.39/bin/ld.gold
> No pkg-config found
> Using runghc version 9.4.2 found on system at:
> /nix/store/8bd87w5v9y7n9b7g3pkzsv344l4zvjwd-ghc-9.4.2/bin/runghc-9.4.2
> Using strip version 2.39 found on system at:
> /nix/store/j061mvdxw70xz20r2r0lr5lk0rcsjf9i-gcc-wrapper-11.3.0/bin/strip
> Using tar found on system at:
> /nix/store/gb4zp7iygbkr3bqfv9p38q3ci04m4m5s-gnutar-1.34/bin/tar
> No uhc found
> *** abort because of serious configure-time warning from Cabal
> ```
> 
> The additional `hspec-core-2.10.6-1YaOpTrOnZw95xaaCyezQr` it complains about doesn't seem to actually exist.

I've hit that while hacking on other things. I believe it's caused by doing overrides of the inputs of one package without changing the versions in the rest of the package set. 

Maybe it's worth looking at the transitive inputs of that derivation to confirm that this isn't a Nix level mistake.

**Comment by lf-** [comment]
re: Servant it's a wip pr at present: https://github.com/haskell-servant/servant/pull/1592

**Comment by sternenseemann** [comment]
`hspec-core` is overridden globally (and nowhere else it seems) and the alleged `hspec-core-2.10.6-1YaOpTrOnZw95xaaCyezQr` doesn't show up in the dependency graph printed by Cabal itself which makes this extremely strange. No clue where it could be passed in…

I guess worth looking into the tasty overrides for the normal package set, but I doubt it's a Nix mistake tbh.

File: pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/haskell-modules/configuration-ghc-9.4.x.nix]: Done.

**Review by sternenseemann — APPROVED** [review]
I think this is a good basis now, although probably for some of the override a version upgrade may also do the trick at this point. Cleaning up such things would be nice, but is likely not required, since the would need to be cleaned up when we upgrade to 9.2 either way (which is likely not that far away).

**Comment by sternenseemann** [comment]
Happy to review and merge follow up PRs! Limited binary cache for 9.4.2 packages should also be available soon.

**Pull request merged**